### ISSUE TITLE
`discarded_futures`: improper warning

### ISSUE THREAD

**Issue by gmpassos** [dart-lang/linter#4085]
The linter `discarded_futures` is very important but is hard to use if it keeps triggering improper warnings.

```dart
import 'dart:async';

// The code below only tries to show the issue with some improper warnings
// for the linter `discarded_futures`.
void main() {
  var ops = Operations();
  populate(ops);
}

void populate(Operations ops) {
  // The call to `_fooOp1` should NOT trigger a linter warning for
  // `discarded_futures`, since this is just the parameter `function`
  // and it expects a `Function` that returns `FutureOr<int>`.
  // What the `add` method does with the `function` parameter
  // is not a matter outside it, specially if `function` is
  // expected to return `FutureOr`.
  ops.add('foo', function: (a, b) => _fooOp1(a, b, 10));

  // The call to `_fooOp2` should NOT trigger a warning too.
  ops.add('foo', function: (a, b) => _fooOp2(a, b));

  // The case below doesn't trigger a warning since it knows
  // that this is not a call to `_fooOp2` (just passing a `Function`).
  ops.add('foo', function: _fooOp2);
}

class Operations {
  final Map<String, FutureOr<int> Function(int a, int b)> _ops = {};

  void add(String name,
      {required FutureOr<int> Function(int a, int b) function}) {
    _ops[name] = function;
  }
}

FutureOr<int> _fooOp1(int a, int b, int m) {
  if (a >= 0) {
    return (a + b) * m;
  } else {
    return Future.delayed(Duration(seconds: 1), () => (a + b) * m);
  }
}

FutureOr<int> _fooOp2(int a, int b) {
  if (a >= 0) {
    return a + b;
  } else {
    return Future.delayed(Duration(seconds: 1), () => a + b);
  }
}

```

**Comment by gmpassos** [comment]
Fix should be at:
https://github.com/dart-lang/linter/blob/main/lib/src/rules/discarded_futures.dart#L86

**Comment by gmpassos** [comment]
This is passed as a parameter to `add`, and the parameter expects to return a `FutureOr`.

The code below shouldn't trigger a warning:
```
 ops.add('foo', function: (a, b) => _fooOp1(a, b, 10));
```

Since `_fooOp1`  returns a `FutureOr` and the parameter `function` expects that. The returned `FutureOr` of `_fooOp1` is the returned value. It's not a case were a `Future` or `FutureOr` is being discarded, it's being returned.

**Comment by westito** [comment]
@gmpassos You're right, and not. In your case it is clearly a false positive, but if you put the `_fooOp1` in a try-catch block the problem will be visible. The function is not interpreted in-place so the try-catch will be silently ignored. May Linter could check for this, but there can be a try-catch anywhere in the call stack that Linter can't check statically. So it is easier to lint in this case, than not. 
On the other hand, I tried to add `async` to the lambda and the lint was gone, however the problem still exists. GCM's `prefer-return-await` rule lints in this case, but Dart linter not warns about the issue I described. 

I think simply the rule get a wrong name. It is called `discarded_futures`, however actually it isn't checking for discarding, rather check for async function call in a non-async block.  
If it would be named 'async_call_in_sync_block' than you wouldn't say it is a false positive 😄

**Comment by gmpassos** [comment]
The way that "discarded_futures" is implemented should not define what we can expect from it. I think that a lint that warns about discarded futures should exists and it's very important.

IMHO, if the current implementation is not matching the definition of the lint it's a bug (not a feature 😂).

**Comment by lrhn** [comment]
If you read the [*description*](https://dart.dev/tools/linter-rules/discarded_futures) of the `discarded_futures` lint, it clearly states what it warns about and that, well, it *really* should have been named `async_call_in_sync_function`.

(It's also a lint that I'd never enable myself, because calling functions returning `Future`s in non-`async` functions is perfectly reasonable in many cases. There is a place for a lint like `unawaited_futures` that applies to non-`async` functions, but this ain't it.)

**Comment by gmpassos** [comment]
Well, the name should be part of the description too (the main part, BTW).

But yes, the current description matches the implementation (thanks to point that).

The point now: should the implementation+description be "updated"?
